osu.Game/Beatmaps/BeatmapDifficultyManager.cs

```diff
@@ -0,0 +1,281 @@
+            return bindable;
+        }
+
+        /// <summary>
+        /// Computes the difficulty defined by a <see cref="DifficultyCacheLookup"/> key, and stores it to the timed cache.
+        /// </summary>
+        /// <param name="key">The <see cref="DifficultyCacheLookup"/> that defines the computation parameters.</param>
+        /// <param name="beatmapInfo">The <see cref="BeatmapInfo"/> to compute the difficulty of.</param>
+        /// <param name="rulesetInfo">The <see cref="RulesetInfo"/> to compute the difficulty with.</param>
+        /// <returns>The <see cref="StarDifficulty"/>.</returns>
+        private StarDifficulty computeDifficulty(in DifficultyCacheLookup key, BeatmapInfo beatmapInfo, RulesetInfo rulesetInfo)
+        {
+            // In the case that the user hasn't given us a ruleset, use the beatmap's default ruleset.
+            rulesetInfo ??= beatmapInfo.Ruleset;
+
+            try
+            {
+                var ruleset = rulesetInfo.CreateInstance();
+                Debug.Assert(ruleset != null);
+
+                var calculator = ruleset.CreateDifficultyCalculator(beatmapManager.GetWorkingBeatmap(beatmapInfo));
+                var attributes = calculator.Calculate(key.Mods);
+
+                return difficultyCache[key] = new StarDifficulty(attributes.StarRating);
+            }
+            catch
+            {
+                return difficultyCache[key] = new StarDifficulty(0);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to retrieve an existing difficulty for the combination.
+        /// </summary>
+        /// <param name="beatmapInfo">The <see cref="BeatmapInfo"/>.</param>
+        /// <param name="rulesetInfo">The <see cref="RulesetInfo"/>.</param>
+        /// <param name="mods">The <see cref="Mod"/>s.</param>
+        /// <param name="existingDifficulty">The existing difficulty value, if present.</param>
+        /// <param name="key">The <see cref="DifficultyCacheLookup"/> key that was used to perform this lookup. This can be further used to query <see cref="computeDifficulty"/>.</param>
+        /// <returns>Whether an existing difficulty was found.</returns>
+        private bool tryGetGetExisting(BeatmapInfo beatmapInfo, RulesetInfo rulesetInfo, IReadOnlyList<Mod> mods, out StarDifficulty existingDifficulty, out DifficultyCacheLookup key)
+        {
+            // In the case that the user hasn't given us a ruleset, use the beatmap's default ruleset.
+            rulesetInfo ??= beatmapInfo.Ruleset;
+
+            // Difficulty can only be computed if the beatmap and ruleset are locally available.
+            if (beatmapInfo.ID == 0 || rulesetInfo.ID == null)
+            {
+                // If not, fall back to the existing star difficulty (e.g. from an online source).
+                existingDifficulty = new StarDifficulty(beatmapInfo.StarDifficulty);
+                key = default;
```

Comment:
I don't think I can fabricate a scenario to confirm my suspicions for sure right now, but this seems unsafe. If I'm compiling in my head correctly, `default` is `{BeatmapId = 0, RulesetId = 0, Mods = null}`; callers of this method use the value of `key` without checks for `default`, so there's a potential for garbage stores to the cache - which in itself wouldn't be a problem, but the garbage reads that would be possible later on (if I'm not misreading) due to reusing that same `default` key is.

The returned `key` should be probably nullable or something.

Reply:
This method is private, so I'm not concerned about callers potentially doing something like:
```
tryGetExisting(... out key)

// Assume the key hasn't existed (ignore return value)
cache[key] = ...
```

Reply:
Hmm, I suppose. I still would have probably restructured this so that this mistake is impossible to make type-wise but I guess as long is it private...